### PR TITLE
Optimize MiniCPM-o 4.5 Whisper chunk-attention mask construction

### PR DESCRIPTION
**Pull request by frank-2077**
## Purpose

This addresses the bounded MiniCPM-o 4.5 audio-encoder A.1 task in #5069.

### Problem

Before each Whisper/APM encoder invocation, MiniCPM-o 4.5 builds a dense `[B, 1, T, T]` attention mask by combining a padding mask with a chunk-attention mask. The padding path is already tensorized, but the chunk mask is populated row by row in Python:

```python
ret = torch.zeros(size, size, device=device, dtype=torch.bool)
for i in range(size):
    ...
    ret[i, start:ending] = True
```

When `ret` is on CUDA, every row assignment enqueues a small fill kernel. For a 30-second, 16 kHz input, Whisper produces approximately

```text
30 s * 16000 samples/s / 160 samples/Mel-frame / 2 CNN downsampling = 1500 encoder positions
```

so the loop enqueues approximately 1,500 row-fill kernels for one mask build.

Baseline Perfetto trace (`T=1500`):

image

The trace shows the repeated `FillFunctor<bool>` kernels from the row assignments. Trace durations include profiler overhead; the performance tables below use CUDA Events.

### Change

Each row boundary depends only on its query index, chunk size, left-context limit, and lookahead. There are no dependencies between rows, so this change computes all row boundaries together and uses broadcasted key-range comparisons:

```python
indices = torch.arange(size, device=device)
chunk_index = indices // chunk_size
chunk_end = torch.clamp(
    (chunk_index + 1) * chunk_size + num_lookhead,
    max=size,
)

mask = indices.unsqueeze(0) < chunk_end.unsqueeze(1)
if num_left_chunks >= 0:
    chunk_start = torch.clamp(
        (chunk_index - num_left_chunks) * chunk_size,
        min=0,
    )
    mask.logical_and_(indices.unsqueeze(0) >= chunk_start.unsqueeze(1))
```

Vectorized Perfetto trace (`T=1500`):

image

This removes the `O(T)` Python loop and per-row fill launches. It does not change the dense mask's `O(T^2)` size, attention semantics, chunk size, or lookahead. It also introduces no mask cache or persistent state. The core `[T, T]` mask is produced by one broadcasted comparison, while the complete implementation still uses a small fixed number of PyTorch tensor operations.

### Scope

- Changes only MiniCPM-o 4.5 `subsequent_chunk_mask` construction.
- Does not change the Whisper/APM attention kernels.
- Does not add encoder caching or change the attention layout.
- Preserves finite left-context and lookahead behavior.

## Test Plan

The benchmark matrix uses deterministic 16 kHz sine-wave inputs at 5, 15, and 30 seconds. Audio-encoder tests cover batch sizes 1, 2, and 4. All values below are medians.

The end-to-end deployment accepts one audio input per prompt and does not include image, video, or mixed audio-vision inputs. `full` mode still executes Thinker, Talker, Flow/DiT, and HiFT. It is limited to batch size 1 because the current single-GPU Talker path supports one request. `skip_mm_profiling` avoids startup OOM from the maximum dummy video input on a 32 GiB GPU; it does not skip real request execution.

**Hardware:** NVIDIA GeForce RTX 5090, 32 GiB, single GPU  
**PyTorch:** 2.11.0+cu130  
**CUDA:** 13.0  
**dtype:** bfloat16  
**Model:** MiniCPM-o 4.5  
**Audio chunk length:** 1 second / 50 post-CNN positions  
**Audio encoder:** 2 warmups + 5 measured iterations  
**Thinker:** 1 warmup + 5 measured iterations, `max_tokens=1`  
**Full:** 1 warmup + 3 measured iterations, `max_tokens=16`

**vLLM Version:** 0.25.0

**vLLM-Omni Commit:** Benchmarks were collected on `ee33954dff27da317be597449a6c1b5a5df4052b`; the PR branch is rebased onto `ad3a162e`, and the target mask implementation is unchanged between those commits.

### Parity

- Exact boolean-mask parity for 216 combinations of `size`, `chunk_size`, `num_left_chunks`, and `num_lookhead`.
- Bitwise-identical audio embeddings for all 9 duration/batch combinations.
- Maximum embedding absolute difference: `0.0`.
- Generated-audio median durations are unchanged in full mode: 2.72 s, 3.00 s, and 2.84 s for 5 s, 15 s, and 30 s inputs, respectively.

### Audio encoder command and script

`python audio_encoder_baseline.py --model /path/to/MiniCPM-o-4_5 --durations 5,15,30 --batch-sizes 1,2,4 --warmup 2 --iterations 5 --output ./encoder-results/before/results.json --embeddings ./encoder-results/before/embeddings.pt --trace ./encoder-results/before/trace.json --trace-duration 30 --trace-batch-size 1
`
[bench_minicpmo_audio_encoder.py](https://github.com/user-attachments/files/30357207/bench_minicpmo_audio_encoder.py)

### Audio encoder and multimodal TTFT

`Mask` and `Encoder` use CUDA Events. TTFT is the Stage 0 serving TTFT including multimodal preprocessing. Peak memory is the PyTorch allocator delta during the isolated audio-encoder call.

| Audio | Batch | T | Mask, before -> after (ms) | Encoder, before -> after (ms) | Serving TTFT, before -> after (ms) | Encoder peak memory, before -> after (MiB) |
|---:|---:|---:|---:|---:|---:|---:|
| 5 s | 1 | 250 | 1.338 -> 0.090 | 7.224 -> 5.918 | 45.214 -> 37.944 | 18.097 -> 18.097 |
| 5 s | 2 | 250 | 1.335 -> 0.086 | 7.121 -> 5.787 | 69.813 -> 70.683 | 36.129 -> 36.129 |
| 5 s | 4 | 250 | 1.340 -> 0.088 | 8.651 -> 7.378 | 65.646 -> 67.188 | 73.367 -> 73.367 |
| 15 s | 1 | 750 | 3.955 -> 0.089 | 10.751 -> 6.851 | 57.703 -> 51.901 | 56.939 -> 56.939 |
| 15 s | 2 | 750 | 3.953 -> 0.088 | 14.559 -> 10.667 | 84.394 -> 77.914 | 112.949 -> 112.949 |
| 15 s | 4 | 750 | 3.954 -> 0.090 | 22.446 -> 18.568 | 89.117 -> 95.583 | 224.212 -> 224.212 |
| 30 s | 1 | 1500 | 7.881 -> 0.089 | 20.324 -> 12.546 | 78.003 -> 66.901 | 116.926 -> 116.926 |
| 30 s | 2 | 1500 | 7.862 -> 0.090 | 29.403 -> 21.595 | 106.000 -> 92.752 | 232.546 -> 232.546 |
| 30 s | 4 | 1500 | 7.865 -> 0.084 | 46.650 -> 38.803 | 144.855 -> 131.913 | 460.695 -> 460.695 |

For the 30-second cases, mask construction is 88-94x faster and total audio-encoder latency is reduced by approximately 7.8 ms, with unchanged peak allocator memory.

For the 30-second, batch-1 case, the five-run ranges were:

| Metric | Before range (ms) | After range (ms) |
|---|---:|---:|
| Mask | 7.853-7.929 | 0.086-0.094 |
| Audio encoder | 20.289-20.374 | 12.505-12.579 |
| Serving TTFT | 74.700-82.206 | 66.094-76.058 |

Short-input and larger-batch TTFT has scheduler noise larger than the local mask saving, so this PR does not claim every end-to-end case improves. The long-input TTFT improvements follow the expected sequence-length-dependent mask saving.

### Full speech pipeline

Full mode executes Thinker, Talker, Flow/DiT, and HiFT at batch size 1. Peak memory is sampled through NVML every 2 ms across all Omni worker processes.

| Audio | Full wall time, before -> after (ms)  |
|---:|---:|
| 5 s | 767.852 -> 781.839 
| 15 s | 833.829 -> 858.662 
| 30 s | 849.570 -> 843.375 

The full pipeline is dominated by Talker and Token2Wav, and the three-run variance is larger than the local mask saving. These results are included as functional validation rather than the primary performance evidence.

### CODE REVIEW

**Comment by amy-why-3459** [comment]
LGTM.

The vectorized `subsequent_chunk_mask` matches the old row-by-row semantics, and the oracle unit test is the right regression coverage for this change. Perf/accuracy evidence in the description is solid for the local mask/encoder win; the full-pipeline caveat looks correct given Talker/Token2Wav dominance.

Optional:
- `subsequent_chunk_mask` doesn't use `self` — consider `@staticmethod` so the test doesn't need `Class.method(None, ...)`.
- If handy, link the exact encoder timing commands/script for easier reproduction.

**Comment by amy-why-3459** [comment]
@Sy0307 @lishunyang12 @linyueqian PTAL

**Review by amy-why-3459 — APPROVED** [review]

**Comment by frank-2077** [comment]
Thanks for the review! I'd prefer to keep the interface unchanged for this PR, as we want to keep the scope limited to mask-construction performance optimization. The benchmark command and script have already been added to the PR description for reproduction.

**Comment by frank-2077** [comment]
@hsliuustc0106 Hi! Just a quick heads-up:  the CI error seems to be unrelated to the PR changes. Thanks for taking a look!
